### PR TITLE
[Scenes] Added the on-off cluster handler for scene EFS

### PR DESCRIPTION
**Pull request by lpbeliveau-silabs**
Adds a default extension field set handler for the on-off cluster, allowing scenes to interact with the on-off cluster when scenes are invoked.

### CODE REVIEW

File: src/app/clusters/scenes-server/SceneTableImpl.cpp [hunk 1]
```diff
@@ -738,7 +738,8 @@ CHIP_ERROR DefaultSceneTableImpl::SceneSaveEFS(SceneTableEntry & scene)
     if (!HandlerListEmpty())
     {
         uint8_t clusterCount = 0;
-        clusterId cArray[kMaxClustersPerScene];
+        // TODO: Use dynamical allocation with emberAFClusterCount to generate a smaller array
+        clusterId cArray[GENERATED_CLUSTER_COUNT];
```
Comment (jmartinez-silabs) [inline src/app/clusters/scenes-server/SceneTableImpl.cpp:742]: So we are not supporting dynamic endpoint. Is there prevention/protection agasint dynamically created endpoints from accessing this array?

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 2]
```diff
@@ -41,6 +41,165 @@ using namespace chip::app::Clusters;
...
+    {
+        app::DataModel::DecodableList<Scenes::Structs::AttributeValuePair::DecodableType> attributeValueList;
+        Scenes::Structs::AttributeValuePair::DecodableType decodePair;
+
+        TLV::TLVReader reader;
+        TLV::TLVType outer;
+
+        size_t attributeCount = 0;
+
+        VerifyOrReturnError(cluster == Id, CHIP_ERROR_INVALID_ARGUMENT);
```
Comment (jepenven-silabs) [inline src/app/clusters/on-off-server/on-off-server.cpp:141]: where is that `Id` coming from shouldn't it be `OnOff::Id` ?

File: src/app/clusters/scenes-server/SceneTableImpl.h [hunk 3]
```diff
@@ -40,8 +40,7 @@ using clusterId = chip::ClusterId;
 class DefaultSceneHandlerImpl : public scenes::SceneHandler
 {
 public:
-    static constexpr uint8_t kMaxValueSize = 4;
-    static constexpr uint8_t kMaxAvPair    = 15;
+    static constexpr uint8_t kMaxAvPair = 15;
```
Comment (jepenven-silabs) [inline src/app/clusters/scenes-server/SceneTableImpl.h:43]: Nit, if this number is present in the spec, please add the reference in comments.

**Review by jepenven-silabs — COMMENTED** [review]
Couple of question and nits to fix

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 4]
```diff
@@ -41,6 +41,165 @@ using namespace chip::app::Clusters;
...
+    }
+
+    /// @brief Serialize the Cluster's EFS value
+    /// @param endpoint target endpoint
+    /// @param cluster  target cluster
+    /// @param serialisedBytes data to serialize into EFS
+    /// @return CHIP_NO_ERROR if successfully serialized the data, CHIP_ERROR_INVALID_ARGUMENT otherwise
+    CHIP_ERROR SerializeSave(EndpointId endpoint, ClusterId cluster, MutableByteSpan & serialisedBytes) override
+    {
+        app::DataModel::List<const Scenes::Structs::AttributeValuePair::Type> OOattributeValueList;
```
[the remark this answers is not in the document]
Comment (lpbeliveau-silabs) [reply src/app/clusters/on-off-server/on-off-server.cpp:93]: Those are two uppercase "o" meant to represent OnOff

File: src/app/clusters/on-off-server/on-off-server.cpp (hunk 2 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/on-off-server/on-off-server.cpp:141]: Above states Using name space OnOff, but at this point either I pu "Id" everywhere or I put OnOff::Id everywhere.

File: src/app/clusters/scenes-server/SceneTableImpl.h (hunk 3 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/scenes-server/SceneTableImpl.h:43]: It's not in the spec, what about a CHIP_CONFIG?

Comment (jepenven-silabs) [reply src/app/clusters/scenes-server/SceneTableImpl.h:43]: that could do.

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 5]
```diff
@@ -41,6 +41,165 @@ using namespace chip::app::Clusters;
...
+private:
+    /**
+     * @brief Configures EnventControl callback when setting On Off through scenes callback
+     *
+     * @param[in] endpoint endpoint to start timer for
+     * @return EmberEventControl* configured event control
+     */
+    EmberEventControl * sceneEventControl(EndpointId endpoint)
+    {
+        EmberEventControl * controller = OnOffServer::Instance().getEventControl(endpoint);
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:185]: This is going to wipe out any existing timing going on for OnWithTimedOff, right?  Why is that OK?

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 6]
```diff
@@ -41,6 +41,165 @@ using namespace chip::app::Clusters;
...
+}
+#endif // EMBER_AF_PLUGIN_LEVEL_CONTROL
+
+#ifdef EMBER_AF_PLUGIN_SCENES
+static void sceneOnOffCallback(EndpointId endpoint);
+
+class DefaultOnOffSceneHandler : public scenes::DefaultSceneHandlerImpl
+{
+public:
+    bool mSceneOnOffState = false;
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:62]: This does not make sense to me.  This state needs to be per-endpoint, no?

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 7]
```diff
@@ -41,6 +41,165 @@ using namespace chip::app::Clusters;
...
+    // As per spec, 1 attribute is scenable in the on off cluster
+    static constexpr uint8_t scenableAttributeCount = 1;
+
+    DefaultOnOffSceneHandler() = default;
+    ~DefaultOnOffSceneHandler() override {}
+
+    // Default function for OnOff cluster, only puts the OnOff cluster ID in the span if supported on the caller endpoint
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:69]:
```suggestion
    // Default function for OnOff cluster, only puts the OnOff cluster ID in the span if supported on the given endpoint
```

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 8]
```diff
@@ -41,6 +41,165 @@ using namespace chip::app::Clusters;
...
+        app::DataModel::List<const Scenes::Structs::AttributeValuePair::Type> OOattributeValueList;
+        Scenes::Structs::AttributeValuePair::Type OOPairs[scenableAttributeCount];
+
+        bool currentValue;
+
+        // read current on/off value
+        EmberAfStatus status = Attributes::OnOff::Get(endpoint, &currentValue);
+        if (status != EMBER_ZCL_STATUS_SUCCESS)
+        {
+            emberAfOnOffClusterPrintln("ERR: reading on/off %x", status);
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:102]: Please use ChipLog*.

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 9]
```diff
@@ -41,6 +41,165 @@ using namespace chip::app::Clusters;
...
+            return CHIP_ERROR_READ_FAILED;
+        }
+
+        OOPairs[0].attributeID.SetValue(Attributes::OnOff::Id);
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:106]: Wait, why is attributeID optional in AttributeValuePair?  It shouldn't be optional!  Followup to fix that, please.

File: src/app/clusters/on-off-server/on-off-server.cpp (hunk 4 above)
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:93]: I strongly recommend declaring it where it's used, like so:

```
using AttributeValuePair = Scenes::Structs::AttributeValuePair::Type;
```
(at top of this function, so there is less typing) and then fill in OOPairs and:

```
app::DataModel::List<const AttributeValuePair> OOattributeValueList(OOPairs);
```

That said, I agree with the naming comment from @jepenven-silabs : we know this is all "on/off", so why do we need the "OO" prefixes on everything?

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 10]
```diff
@@ -41,6 +41,165 @@ using namespace chip::app::Clusters;
...
+        OOPairs[0].attributeValue = currentValue;
+        OOattributeValueList      = OOPairs;
+
+        TLV::TLVWriter writer;
+        TLV::TLVType outer;
+        // Serialize Extension Field sets in a way consistent with the default Deserialize method
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:112]: Which default Deserialize method?  This should probably point more specifically at the thing we are trying to match.

But more importantly, if the idea is that we have to match the way the scene table does something, shouldn't the scene table just expose a helper for doing that thing?

But my understanding was that we don't _have_ to match that thing.  Is that understanding incorrect?

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 11]
```diff
@@ -41,6 +41,165 @@ using namespace chip::app::Clusters;
...
+        ReturnErrorOnFailure(reader.Next(
+            TLV::kTLVType_Array, TLV::ContextTag(app::Clusters::Scenes::Structs::ExtensionFieldSet::Fields::kAttributeValueList)));
+        ReturnErrorOnFailure(attributeValueList.Decode(reader));
+
+        ReturnErrorOnFailure(attributeValueList.ComputeSize(&attributeCount));
+        VerifyOrReturnError(attributeCount <= scenableAttributeCount, CHIP_ERROR_BUFFER_TOO_SMALL);
+        auto pair_iterator = attributeValueList.begin();
+
+        // Assumes the OnOff cluster has only one Attribute
+        pair_iterator.Next();
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:155]: This needs to handle invalid TLV, no?  So check whether Next() returned true, check GetStatus(), etc.

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 12]
```diff
@@ -41,6 +41,165 @@ using namespace chip::app::Clusters;
...
+        if (decodePair.attributeID.HasValue())
+        {
+            // If attribute ID was encoded, verify it is the proper ID for the OnOff attribute
+            VerifyOrReturnError(decodePair.attributeID.Value() == Attributes::OnOff::Id, CHIP_ERROR_INVALID_ARGUMENT);
+        }
+
+        mSceneOnOffState = decodePair.attributeValue;
+
+        // Verify that the EFS was completely read
+        ReturnErrorOnFailure(pair_iterator.GetStatus());
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:168]: This needs to be _before_ we use GetValue(), effectively.  So as soon as Next() returns false.

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 13]
```diff
@@ -41,6 +41,165 @@ using namespace chip::app::Clusters;
...
+        ReturnErrorOnFailure(reader.ExitContainer(outer));
+
+        OnOffServer::Instance().scheduleTimerCallbackMs(sceneEventControl(endpoint), timeMs);
+
+        return CHIP_NO_ERROR;
+    }
+
+private:
+    /**
+     * @brief Configures EnventControl callback when setting On Off through scenes callback
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:178]: What is Envent?

File: src/app/clusters/scenes-server/SceneTableImpl.cpp (hunk 1 above)
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/SceneTableImpl.cpp:742]: Why these changes?  These have nothing to do with on/off.  They should be in a separate PR with a separate clear description of what they are aiming at.

File: src/app/clusters/scenes-server/scenes-server.cpp [hunk 14]
```diff
@@ -762,8 +773,7 @@ void ScenesServer::HandleCopyScene(HandlerContext & ctx, const Commands::CopySce
         AddResponseOnError(ctx, response, Attributes::SceneTableSize::Get(ctx.mRequestPath.mEndpointId, &endpointTableSize)));
 
     // Get Scene Table Instance
-    scenes::SceneTable<scenes::ExtensionFieldSetsImpl> * sceneTable =
-        scenes::GetSceneTableImpl(ctx.mRequestPath.mEndpointId, endpointTableSize);
+    SceneTable * sceneTable = scenes::GetSceneTableImpl(ctx.mRequestPath.mEndpointId, endpointTableSize);
```
Comment (bzbarsky-apple) [inline src/app/clusters/scenes-server/scenes-server.cpp:776]: All these cleanup bits should _also_ not be in the same PR as substantive changes.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/lighting-app/silabs/src/AppTask.cpp [hunk 15]
```diff
@@ -86,6 +86,9 @@ CHIP_ERROR AppTask::Init()
         appError(err);
     }
 
+    // Registers Scene handlers for the On/Off cluster on the server
+    app::Clusters::Scenes::ScenesServer::Instance().RegisterSceneHandler(OnOffServer::Instance().GetSceneHandler());
```
[the remark this answers is not in the document]
Comment (lpbeliveau-silabs) [reply examples/lighting-app/silabs/src/AppTask.cpp:90]: Because the app layer should be able to implement its own handlers, say it would like to implement one single handler for all clusters. The idea was the there needs to be a way for the user to decide wether to use the default handlers or not and that should not involve having to modify the cluster code. I could add a config : CHIP_USE_DEFAULT_SCENE_HANDLERS but wouldn't that be less explicit?

File: src/app/clusters/on-off-server/on-off-server.cpp (hunk 9 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/on-off-server/on-off-server.cpp:106]: Issue [here](https://github.com/project-chip/connectedhomeip/issues/27076)

File: src/app/clusters/on-off-server/on-off-server.cpp (hunk 10 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/on-off-server/on-off-server.cpp:112]: The default is the one in SceneTableImpl.h, this will be added.

Considering the matching, a handler needs to be consistent with itself, so that a serialize from AddScene (SerializeAdd) and one from StoreScene (SerializeSave) store the data the same way and is retrievable by both Deserialize and ApplyScene methods.

If the user wants to store in a different way, he has to override the default Deserialize and SerializeAdd.

File: src/app/clusters/on-off-server/on-off-server.cpp (hunk 5 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/on-off-server/on-off-server.cpp:185]: Isn't that the desired behavior? Say I change scene and user presses a button toggling onOff, it will override the change scene, and vice-versa, if user presses a button and change scenes happen in between that press and the end of the transition time. 

Do we want to queue different event every-time a new scene is invoke? Say if a user calls recall scene on scene1 with a transition time of 30s, and then changes his mind and calls recall scene2 with a transition of 10s, then the scene would transition to scene2, then scene1. Similar behavior could happen between user action and scene recall. Say user presses button, which starts a transition on-off -> 1, then recall scene2 which has on-off = 0 and a transition time of 0.

Scene2 makes on-off=0, then the button press happens and cancels user option. This doesn't seem like the desired behavior.

File: src/app/clusters/scenes-server/SceneTableImpl.cpp (hunk 1 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/scenes-server/SceneTableImpl.cpp:742]: Because the number of clusters in an endpoint is greater than the maximum clusters allowed per scenes, which means if we don't iterate over all the clusters of an endpoint, it is possible to not hit it's sceneable ones (which is what happened during tests)

This will be moved to a separate PR.

File: examples/lighting-app/silabs/src/AppTask.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply examples/lighting-app/silabs/src/AppTask.cpp:90]: I think ideally things would "work out of the box", and then the app can customize them as needed.  The need to make this `RegisterSceneHandler` call for Scenes to work at all for On/Off is _very_ non-obvious.

So I think the right thing to do is to have On/Off handle this by default but have a compile-time or run-time option to turn that off for applications that want to control things more directly.

File: src/app/clusters/on-off-server/on-off-server.cpp (hunk 5 above)
Comment (bzbarsky-apple) [reply src/app/clusters/on-off-server/on-off-server.cpp:185]: Let me ask a more direct question:  If I do OnWithTimedOff, and then I change scene, should that cancel the timed off?  The timed off bits are generally meant to be used by sensor-based lighting setups, I believe, and I would have thought that you might want that even if you have scenes going on.  @Rob-Houtepen-Signify thoughts?

File: src/app/clusters/on-off-server/on-off-server.cpp (hunk 10 above)
Comment (bzbarsky-apple) [reply src/app/clusters/on-off-server/on-off-server.cpp:112]: My main point is that we should ideally not have duplicated code that is supposed to stay in sync.

If we need to be in sync with something, we should share the relevant code.  If we don't need to stay in sync with anything, then we shouldn't say we need to be in sync with it.

File: src/app/clusters/scenes-server/scenes-server.cpp (hunk 14 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/scenes-server/scenes-server.cpp:776]: Pr created [here](https://github.com/project-chip/connectedhomeip/pull/27079)

File: src/app/clusters/on-off-server/on-off-server.cpp (hunk 6 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/on-off-server/on-off-server.cpp]: Now handled per-endpoint using EndpointStatePair and StatePairBuffer structs.

File: src/app/clusters/on-off-server/on-off-server.cpp (hunk 10 above)
Comment (lpbeliveau-silabs) [reply src/app/clusters/on-off-server/on-off-server.cpp]: PR [here](https://github.com/project-chip/connectedhomeip/pull/27089) to introduce helper functions.

File: src/app/clusters/on-off-server/on-off-server.cpp (hunk 5 above)
Comment (Rob-Houtepen-Signify) [reply src/app/clusters/on-off-server/on-off-server.cpp:271]: The OnWithTimedOff related use-case is indeed about sensor based for automatic control, but allowing for manual overwrites. Originally, people where more thinking about On/Off commands as result of a manual override, but like you indicated, a Recall Scene command may also be the result of a manual override.

The spec text below the state diagram says:
Off1 : Any command which causes the OnOff attribute to be set to FALSE, e.g. Off, Toggle or OffWithEffect.
On2 : Any command which causes the OnOff attribute to be set to TRUE, e.g. On, Toggle or OnWithRecallGlobalScene.

So on a Recall Scene I would check whether the resulting state for the OnOff attribute is True or False, and treat this as an internal On of Off command. (Even if the OnOff attribute does not change, still treat it as internal On or Off command to have the effect of the manual override.)

Comment (bzbarsky-apple) [reply src/app/clusters/on-off-server/on-off-server.cpp:271]: @Rob-Houtepen-Signify The real issue I am having is that this proposed code is reusing the timer that drives the OnTime/OffWaitTime updates to time the scene transition.

So for example, if we are in "Delayed Off" state and we get a scene change with OnOff = false, per spec it seems like it should leave us in that state and keep decrementing OffWaitTime.  But as far as I can tell it will instead stop updating OffWaitTime, and probably leave us stuck in "Delayed Off".  If a lot of time passes, the next OnWithTimedOff command, will start up the timer again but stay in "Delayed Off" instead of transitioning to "Timed On" as it should....

The actual application of the scene does in fact pretend to be an On or Off command, so that part is fine.

@lpbeliveau-silabs am I missing something in the analysis above?

Comment (lpbeliveau-silabs) [reply src/app/clusters/on-off-server/on-off-server.cpp:271]: @bzbarsky-apple That is correct, we will indeed restart from where we were at once the next function is called in some cases. 
What I propose is to add:
`uint16_t onTime                     = 0;
uint16_t offWaitTime                = 0;  
OnOff::Attributes::OffWaitTime::Set(endpoint, &offWaitTime);
OnOff::Attributes::OnTime::Set(endpoint, &onTime);`
(In a proper styling )To the original sceneOnOffCallback() below.

Comment (bzbarsky-apple) [reply src/app/clusters/on-off-server/on-off-server.cpp:271]: I don't think resetting OnTime and OffWaitTime on scene change would match the spec.

If the idea is that scene change is supposed to act as if a On or Off was sent, then per "1.5.8. State Description" in appclusters:

1. Changing scene to OnOff = false while in "Delayed Off" state should not interfere with the OffWaitTime countdown.
2. Changing scene to OnOff = true while in "Timed On" state should not interfere with the OnTime countdown.

Or at least that's my understanding of the spec and of @Rob-Houtepen-Signify's comments above.

Comment (lpbeliveau-silabs) [reply src/app/clusters/on-off-server/on-off-server.cpp]: Current version now has a sceneHandlerEvenControl array for each endpoint

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 16]
```diff
@@ -645,28 +883,31 @@ bool OnOffServer::areStartUpOnOffServerAttributesNonVolatile(EndpointId endpoint
...
  * @brief event control object for an endpoint
  *
- * @param[in] endpoint
+ * @param[in] endpoint target endpoint
+ * @param[in] eventControlArray Array where to find the event control
+ * @param[in] eventControlArraySize Size of the event control array
  * @return EmberEventControl* configured event control
  */
-EmberEventControl * OnOffServer::getEventControl(EndpointId endpoint)
+EmberEventControl * OnOffServer::getEventControl(chip::EndpointId endpoint, EmberEventControl * eventControlArray,
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:891]: Might have been nicer with `const Span<EmberEventControl> &` as the second arg and no separate size arg.  But maybe not, if the Span constructor that takes a C array is "explicit"...  Either way.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 17]
```diff
@@ -41,15 +41,249 @@ using namespace chip::app::Clusters;
...
+
+        /// @brief Removes Pair and decrements Pair count if the endpoint existed in the array
+        /// @param endpoint : endpoint id of the pair
+        CHIP_ERROR RemovePair(const EndpointId endpoint)
+        {
+            uint16_t position;
+            VerifyOrReturnValue(CHIP_NO_ERROR == FindPair(endpoint, position), CHIP_NO_ERROR);
+
+            uint16_t nextPos = static_cast<uint16_t>(position + 1);
+            uint16_t moveNum = static_cast<uint16_t>(MAX_ENDPOINT_COUNT - nextPos);
```
[the remark this answers is not in the document]
Comment (lpbeliveau-silabs) [reply src/app/clusters/on-off-server/on-off-server.cpp:135]: Should have been `kOnOffMaxEnpointCount`, I used to define my array size `MAX_ENDPOINT_COUNT` but then I figured it should never be above `kOnOffMaxEnpointCount` , so `ArraySize(mStatePairBuffer) - nextPos` would have been the logic I was going for. However  this was totally moving garbage around though unless the array is full so this `mPairCount` is definetely better.

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 18]
```diff
@@ -645,28 +887,31 @@ bool OnOffServer::areStartUpOnOffServerAttributesNonVolatile(EndpointId endpoint
...
 {
     uint16_t index = emberAfGetClusterServerEndpointIndex(endpoint, OnOff::Id, EMBER_AF_ON_OFF_CLUSTER_SERVER_ENDPOINT_COUNT);
-    if (index >= ArraySize(gEventControls))
+    if (index >= eventControlArray.size())
     {
         return nullptr;
     }
-    return &gEventControls[index];
+
+    return &eventControlArray.data()[index];
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:903]:
```suggestion
    return &eventControlArray[index];
```

File: src/app/clusters/on-off-server/on-off-server.cpp [hunk 19]
```diff
@@ -645,28 +887,31 @@ bool OnOffServer::areStartUpOnOffServerAttributesNonVolatile(EndpointId endpoint
...
  * @brief event control object for an endpoint
  *
- * @param[in] endpoint
+ * @param[in] endpoint target endpoint
+ * @param[in] eventControlArray Array where to find the event control
+ * @param[in] eventControlArraySize Size of the event control array
  * @return EmberEventControl* configured event control
  */
-EmberEventControl * OnOffServer::getEventControl(EndpointId endpoint)
+EmberEventControl * OnOffServer::getEventControl(chip::EndpointId endpoint, const Span<EmberEventControl> & eventControlArray)
```
Comment (bzbarsky-apple) [inline src/app/clusters/on-off-server/on-off-server.cpp:895]:
```suggestion
EmberEventControl * OnOffServer::getEventControl(EndpointId endpoint, const Span<EmberEventControl> & eventControlArray)
```

**Review by jmartinez-silabs — APPROVED** [review]
